File: LeetCode-C++/LeetCode-C++/二叉树/_889_根据前序和后序遍历构造二叉树.cpp
//
//  _889_根据前序和后序遍历构造二叉树.cpp
//  LeetCode-C++
//
//  Created by warden on 2019/12/12.
//  Copyright © 2019 warden. All rights reserved.
//
// https://leetcode-cn.com/problems/construct-binary-tree-from-preorder-and-postorder-traversal/

#include <stdio.h>
#include <vector>
#include <stack>
#include "TreeNode.hpp"

using namespace std;


class Solution {
public:
    TreeNode* constructFromPrePost(vector<int>& pre, vector<int>& post) {
        if (pre.empty()) {
            return NULL;
        }
        TreeNode *root = new TreeNode(pre[0]);
        stack<TreeNode*> st;
        st.push(root);
        for (int i = 1, j = 0; i < pre.size(); i++) {
            TreeNode *node = new TreeNode(pre[i]);
            while (st.top()->val == post[j]) {
                st.pop();
                j++;
            }
            if (!st.top()->left) {
                st.top()->left = node;
            } else {
                st.top()->right = node;
            }
            st.push(node);
        }
        return root;
    }
};

/**
 示例：

 输入：pre = [1,2,4,5,3,6,7], post = [4,5,2,6,7,3,1]
 输出：[1,2,3,4,5,6,7]
 */
/**
int main(int argc, const char * argv[]) {
    vector<int> preorder = {1,2,4,5,3,6,7};
    vector<int> inorder = {4,5,2,6,7,3,1};
    Solution* solution = new Solution();
    TreeNode* root = solution->constructFromPrePost(preorder, inorder);
    printTreeByLevel(root);
    return 0;
}
 */

File: LeetCode-Swift/LeetCode-Swift/二叉树/_889_根据前序和后序遍历构造二叉树.swift
//
//  _889_根据前序和后序遍历构造二叉树.swift
//  LeetCode-Swift
//
//  Created by warden on 2019/12/12.
//  Copyright © 2019 warden. All rights reserved.
//
// https://leetcode-cn.com/problems/construct-binary-tree-from-preorder-and-postorder-traversal/

import Foundation

class _889_根据前序和后序遍历构造二叉树 {
    func constructFromPrePost(_ pre: [Int], _ post: [Int]) -> TreeNode? {
        guard !pre.isEmpty else { return nil }
        let root = TreeNode(pre.first!)
        var nodes = [root]
        var j = 0
        for i in 1..<pre.count {
            let curNode = TreeNode(pre[i])
            while nodes.last!.val == post[j] {
                nodes.removeLast()
                j += 1
            }
            if nodes.last?.left == nil {
                nodes.last?.left = curNode
            } else {
                nodes.last?.right = curNode
            }
            nodes.append(curNode)
        }
        return root
    }
}

/**
 输入：pre = [1,2,4,5,3,6,7], post = [4,5,2,6,7,3,1]
 输出：[1,2,3,4,5,6,7]
 */
func test_889() {
    let solution = _889_根据前序和后序遍历构造二叉树()
    let root = solution.constructFromPrePost([1, 2, 4, 5, 3, 6, 7], [4, 5, 2, 6, 7, 3, 1])
    print(root!)
}
